components/camel-tahu/src/main/java/org/apache/camel/component/tahu/TahuEdgeClientCallback.java

```diff
@@ -0,0 +1,236 @@
+                }
+                return;
+            }
+
+            if (!topic.isType(MessageType.NCMD) && !topic.isType(MessageType.DCMD)) {
+                LOG.debug(loggingMarker, "Received unexpected Sparkplug message of type {} - ignoring", topic.getType());
+                return;
+            }
+
+            try {
+                PayloadDecoder<SparkplugBPayload> decoder = new SparkplugBPayloadDecoder();
+                payload = decoder.buildFromByteArray(mqttMessage.getPayload(), null);
+
+                if (topic.isType(MessageType.NCMD)) {
+                    handleNCMDMessage(payload);
+                } else if (topic.isType(MessageType.DCMD)) {
+                    handleDCMDMessage(payload, topic.getDeviceId());
+                }
+            } catch (Exception e) {
+                LOG.error(loggingMarker, "Exception caught decoding Sparkplug message with topic {} and payload {}", topic,
+                        new String(mqttMessage.getPayload()));
+                throw new RuntimeCamelException(e);
+            }
+        } finally {
+            LOG.trace(loggingMarker, "{}: ClientCallback messageArrived complete", mqttClientId);
+        }
+    }
+
+    void handleNCMDMessage(SparkplugBPayload ncmdPayload) {
+        List<Metric> responseMetrics = tahuEdgeNodeMetricHandler.processCMDMetrics(ncmdPayload, edgeNodeDescriptor);
+
+        if (responseMetrics.isEmpty()) {
+            LOG.warn(loggingMarker, "Received NCMD with no valid metrics to write for {} - ignoring",
+                    edgeNodeDescriptor);
+            return;
+        }
+
+        SparkplugBPayload ndataPayload = new SparkplugBPayload.SparkplugBPayloadBuilder()
+                .addMetrics(responseMetrics)
+                .createPayload();
+
+        LOG.debug(loggingMarker, "Publishing NDATA based on NCMD message for {}", edgeNodeDescriptor);
+
+        client.publishNodeData(ndataPayload);
+    }
+
+    void handleDCMDMessage(SparkplugBPayload dcmdPayload, String deviceId) {
+        DeviceDescriptor deviceDescriptor = new DeviceDescriptor(edgeNodeDescriptor, deviceId);
+
+        List<Metric> responseMetrics = tahuEdgeNodeMetricHandler.processCMDMetrics(dcmdPayload, deviceDescriptor);
+
+        if (responseMetrics.isEmpty()) {
+            LOG.warn(loggingMarker, "Received DCMD with no valid metrics to write for {} - ignoring", deviceDescriptor);
+            return;
+        }
+
+        SparkplugBPayload ddataPayload = new SparkplugBPayload.SparkplugBPayloadBuilder()
+                .addMetrics(responseMetrics)
+                .createPayload();
+
+        LOG.debug(loggingMarker, "Publishing DDATA based on DCMD message for {}", deviceDescriptor);
+
+        client.publishDeviceData(deviceId, ddataPayload);
+    }
+
+    @Override
+    public void shutdown() {
+        LOG.trace(loggingMarker, "ClientCallback shutdown called");
+
+        LOG.trace(loggingMarker, "ClientCallback shutdown complete");
```

Comment:
Are these needed?

Reply:
No, not really. I'll remove them throughout.